yarn-project/noir-libs/value-note/src/utils.nr

```diff
@@ -1,184 +1,125 @@
 use dep::aztec::oracle::get_public_key::get_public_key;
 use dep::aztec::state_vars::set::Set;
-use dep::aztec::types::point::Point;
 use crate::{
     filter::filter_notes_min_sum,
     value_note::{ValueNote, VALUE_NOTE_LEN},
 };
 
-fn spend_notes(
+// Sort the note values (0th field) in descending order.
+// Pick the fewest notes whose sum is equal to or greater than `amount`.
+fn create_note_getter_options_for_decreasing_balance(amount: Field) -> NoteGetterOptions<ValueNote, VALUE_NOTE_LEN, Field> {
+    NoteGetterOptions::with_filter(filter_notes_min_sum, amount).sort(0, SortOrder.DESC)
+}
+
+// Creates a new note for the recipient.
+// Inserts it to the recipient's balance.
+fn increase_balance(
     context: &mut PrivateContext,
     balance: Set<ValueNote, VALUE_NOTE_LEN>,
     amount: Field,
-    owner: Field,
+    recipient: Field,
 ) {
-    let options = NoteGetterOptions::with_filter(filter_notes_min_sum, amount);
-    let opt_notes = balance.get_notes(context, options);
-
-    let mut sum = 0;
-    for i in 0..opt_notes.len() {
-        if opt_notes[i].is_some() {
-            let note = opt_notes[i].unwrap_unchecked();
-
-            // Ensure the notes are actually owned by the owner (to prevent user from generating a valid proof while
-            // spending someone else's notes).
-            assert(note.owner == owner);
-
-            // Removes the note from the owner's set of notes.
-            balance.remove(context, note);
-
-            sum += note.value;
-        }
-    }
-
-    assert(sum as u120 >= amount as u120);
-
-    // Creates change note for the owner.
-    let change_value = sum - amount;
-    let mut change_note = ValueNote::new(change_value, owner);
-    balance.insert(context, &mut change_note);
-
-    // Emit the newly created encrypted note preimages via oracle calls.
-    let mut encrypted_data = [0; VALUE_NOTE_LEN];
-    if change_value != 0 {
-        encrypted_data = change_note.serialise();
-    };
+    let mut note = ValueNote::new(amount, recipient);
+    create_note(context, balance, recipient, &mut note);
 
-    let encryption_pub_key = get_public_key(owner);
-    emit_encrypted_log(
-        context,
-        (*context).this_address(),
-        balance.storage_slot,
-        encryption_pub_key,
-        encrypted_data,
-    );
+    // It won't compile if Set.insert() is in an if statement :(
+    // if amount as u120 > 0 {
+    //     create_note(context, balance, recipient, &mut note);
+    // }
 }
 
-/*
-  Spends one note from a set of 4 notes.
-  Details: Reads 4 notes from the user's `balance`: [n_{o}, n_{o+1}, n_{o+2}, n_{o+3}]
-  where "o" is the note_offset. Then, spends the note: [n_{o + i}] 
-  where "i" is the spend_note_index. This gives more control to the user on which
-  note is to be spent. 
-*/
-fn spend_one_note(
+// Find some notes that add up to the amount.
+// Removes those note from owner's balance.
+// Fail if the sum of the selected notes is less than the amount.
```

Comment:
```suggestion
// Find some of the `owner`'s notes whose values add up to the `amount`.
// Remove those notes.
// If the value of the removed notes exceeds the requested `amount`, create a new note containing the excess value, so that exactly `amount` is removed.
// Fail if the sum of the selected notes is less than the amount.
```